apps/consulting/posts/scaling-python-banking-edition.md

```diff
@@ -0,0 +1,677 @@
+alt="A picture of a jupyter notebook input and output cells, next to an image of the cell contents equivilent dask graph being run"
+width="700px"
+/>
+
+**_!!!! (We need to walk the reader through the above figure) !!!!_**
+
+However, as we noted earlier, our resources may be limited. What
+happens if our dataframe is larger than the cluster itself?
+
+### Dataframe Group-by
+
+As it turns out, we ran into this issue with some of our larger
+aggregations. We were finding that when we submitted jobs, workers
+would be forced to pause or restart when the memory thresholds for the
+cluster were reached, in some cases causing the job to fail.
+
+```python
+>>> graph = build_graph(paths_10)
+>>> fut = client.compute(graph)
+>>> fut.result().info()
+
+2023-03-31 18:58:33,630 - distributed.worker_memory - WARNING - Worker is at 83% memory usage. Pausing worker.  Process memory: 3.33 GiB -- Worker memory limit: 4.00 GiB
+2023-03-31 18:58:33,792 - distributed.worker_memory - WARNING - Worker is at 76% memory usage. Resuming worker. Process memory: 3.07 GiB -- Worker memory limit: 4.00 GiB
+2023-03-31 18:58:33,893 - distributed.worker_memory - WARNING - Worker is at 89% memory usage. Pausing worker.  Process memory: 3.60 GiB -- Worker memory limit: 4.00 GiB
+2023-03-31 18:58:33,934 - distributed.worker_memory - WARNING - Worker exceeded 95% memory budget. Restarting
+2023-03-31 18:58:33,977 - distributed.nanny - WARNING - Restarting worker
+2023-03-31 18:58:51,739 - distributed.worker_memory - WARNING - Worker is at 86% memory usage. Pausing worker.  Process memory: 3.46 GiB -- Worker memory limit: 4.00 GiB
+2023-03-31 18:58:52,334 - distributed.worker_memory - WARNING - Worker exceeded 95% memory budget. Restarting
+2023-03-31 18:58:52,376 - distributed.nanny - WARNING - Restarting worker
+---------------------------------------------------------------------------
+KilledWorker                              Traceback (most recent call last)
+File <timed exec>:3
+
+File ~/.conda/envs/bank/lib/python3.8/site-packages/distributed/client.py:283, in Future.result(self, timeout)
+    281 if self.status == "error":
+    282     typ, exc, tb = result
+--> 283     raise exc.with_traceback(tb)
+    284 elif self.status == "cancelled":
+    285     raise result
+
+KilledWorker: ("('dataframe-groupby-sum-chunk-d4e753da53a3eff842bab7cbba152105-e7aebcddb75c44615a5bb2eb1e6f9433', 82)", <WorkerState 'tcp://127.0.0.1:42571', name: 1, status: closed, memory: 0, processing: 39>)
+```
+
+<img
+src="/posts/scaling-python-banking-edition/dd-larger-than-cluster-dask.png"
+alt="An image of the dask dashboard with the results a dask graph representing the code above and the resulting run failing."
+/>
+
+**_!!!!(Can we add some sort of annotation to the above image to show how a reader could interpret the graph to understand it was an out-of-memory failure? Or is the only way to know this from inspecting the traceback)!!!!_**
```

Comment:
Flagging this.

Reply:
I added more description to the paragraph below. Let me know what you think